src/main/java/org/opensearch/ad/ratelimit/CheckpointReadQueue.java

```diff
@@ -0,0 +1,374 @@
+        return ActionListener.wrap(response -> {
+            final MultiGetItemResponse[] itemResponses = response.getResponses();
+            Map<String, MultiGetItemResponse> successfulRequests = new HashMap<>();
+
+            // lazy init since we don't expect retryable requests to happen often
+            Set<String> retryableRequests = null;
+            Set<String> notFoundModels = null;
+            for (MultiGetItemResponse itemResponse : itemResponses) {
+                String modelId = itemResponse.getId();
+                if (itemResponse.isFailed()) {
+                    final Exception failure = itemResponse.getFailure().getFailure();
+                    if (failure instanceof IndexNotFoundException) {
+                        for (EntityRequest origRequest : toProcess) {
+                            // submit to cold start queue
+                            entityColdStartQueue.put(origRequest);
+                        }
+                        return;
+                    } else if (ExceptionUtil.isNotFound(failure)) {
+                        if (notFoundModels == null) {
+                            notFoundModels = new HashSet<>();
+                        }
+                        notFoundModels.add(modelId);
+                    } else if (ExceptionUtil.isRetryAble(failure)) {
+                        if (retryableRequests == null) {
+                            retryableRequests = new HashSet<>();
+                        }
+                        retryableRequests.add(modelId);
+                    } else {
+                        LOG.info("Unexpected failure", failure);
+                    }
+                } else {
+                    successfulRequests.put(modelId, itemResponse);
+                }
+            }
+
+            // deal with not found model
+            if (notFoundModels != null) {
+                for (EntityRequest origRequest : toProcess) {
+                    Optional<String> modelId = origRequest.getModelId();
+                    if (modelId.isPresent() && notFoundModels.contains(modelId.get())) {
+                        // submit to cold start queue
+                        entityColdStartQueue.put(origRequest);
+                    }
+                }
+            }
+
+            processCheckpointIteration(0, toProcess, successfulRequests, retryableRequests);
+        }, exception -> {
+            if (ExceptionUtil.isOverloaded(exception)) {
+                LOG.error("too many get AD model checkpoint requests or shard not avialble");
+                setCoolDownStart();
+            } else if (ExceptionUtil.isRetryAble(exception)) {
+                // retry all of them
+                super.putAll(toProcess);
+            } else {
+                LOG.error("Fail to restore models", exception);
+            }
+        });
+    }
+
+    private void processCheckpointIteration(
+        int i,
```

Comment:
I think this name be more descriptive.

Reply:
Please check https://google.github.io/styleguide/cppguide.html#General_Naming_Rules

"Note that certain universally-known abbreviations are OK, such as i for an iteration variable and T for a template parameter."